### PR TITLE
Support marshmallow 4.0.0 & add e2e tests

### PR DESCRIPTION
**Pull request by Kraysent**
Hi! I saw your repo in the [StackOverflow answer](https://stackoverflow.com/a/76113256) and wanted to use it, but discovered that it no longer works since the update of `marshmallow` to version 4 (which introduced [a bunch of breaking changes](https://marshmallow.readthedocs.io/en/latest/upgrading.html#upgrading-to-4-0)) since it uses fields that were removed in the update. 

This PR addresses all the necessary changes and also adds some simple e2e tests, as none of the existing tests failed when the `marshmallow` dependency was upgraded in the `requirements.txt`.

Let me know if anything needs to be changed.

### CODE REVIEW

**Comment by netwolfuk** [comment]
This is awesome @Kraysent :heart: 

@daniil-berg Do you think you'd get a chance to review, merge and release a new `marshmallow-generic` version?

File: src/marshmallow_generic/__init__.py [hunk 1]
```diff
@@ -51,7 +51,8 @@
 )
 from marshmallow.exceptions import ValidationError
 from marshmallow.schema import Schema, SchemaOpts
-from marshmallow.utils import EXCLUDE, INCLUDE, RAISE, missing, pprint
+from marshmallow.constants import EXCLUDE, INCLUDE, RAISE, missing
+from pprint import pprint
```
Comment (daniil-berg) [inline src/marshmallow_generic/__init__.py:55]: The `pprint` re-export [was removed in `marshmallow` ](https://github.com/marshmallow-code/marshmallow/commit/b393c1fba7bfc0220637399bd3452d179e808cbd) and can therefore be omitted here as well.

```suggestion
```

File: src/marshmallow_generic/schema.py [hunk 2]
```diff
@@ -52,14 +52,13 @@ class FooSchema(GenericSchema[Foo]):
...
-        partial: Union[bool, Sequence[str], set[str]] = False,
-        unknown: Optional[str] = None,
-        many: bool = False,  # usage discouraged
+        only: types.StrSequenceOrSet | None = None,
+        exclude: types.StrSequenceOrSet = (),
+        many: bool | None = None,
+        load_only: types.StrSequenceOrSet = (),
+        dump_only: types.StrSequenceOrSet = (),
+        partial: bool | types.StrSequenceOrSet | None = None,
+        unknown: types.UnknownOption | None = None,
```
Comment (daniil-berg) [inline src/marshmallow_generic/schema.py:61]: This is a small detail, but I consciously decided to move the `many` parameter to the end of the list (and document it last) because its use in the constructor is heavily discouraged (see documentation). Fortunately, this is not a change that breaks the parent `Schema` contract since all of those must be provided as (non-positional) keyword arguments anyway.
```suggestion
        only: types.StrSequenceOrSet | None = None,
        exclude: types.StrSequenceOrSet = (),
        load_only: types.StrSequenceOrSet = (),
        dump_only: types.StrSequenceOrSet = (),
        partial: bool | types.StrSequenceOrSet | None = None,
        unknown: types.UnknownOption | None = None,
        many: bool | None = None,
```

**Review by daniil-berg — CHANGES_REQUESTED** [review]
I apologize for ignoring this PR for so long! Thank you @Kraysent  for the work you put in and thanks to the others for reacting/commenting here. If you can address a few of those minor issues below, I will happily merge this and release a new version.

The minimum required Python version will have to be bumped to 3.10 as well, but this does not need to be part of this PR.

**Comment by daniil-berg** [comment]
On second thought, I'll just merge this now and make those changes myself. I have stalled this long enough. Thanks again.

**Comment by Kraysent** [comment]
No worries, thanks for this project!
